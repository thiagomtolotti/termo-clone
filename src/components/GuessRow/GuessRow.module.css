.guessRow {
  display: flex;
  gap: 4px;
  margin: 4px;
}

.guessRow.active {
  font-weight: 600;
}
