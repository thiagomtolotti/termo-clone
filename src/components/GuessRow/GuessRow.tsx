"use client";

import { useEffect, useState } from "react";
import { GuessLetter } from "../GuessLetter/GuessLetter";
import styles from "./GuessRow.module.css";

const WORD_SIZE = 5;
const letterRegex = /^[a-zA-Z]$/;

interface GuessRowProps {
  active?: boolean;
}

export const GuessRow = ({ active = false }: GuessRowProps) => {
  const [activeGuessIndex, setActiveGuessIndex] = useState(0);

  useEffect(() => {
    const handleArrowClick = (ev: KeyboardEvent) => {
      if (ev.key === "ArrowRight") {
        setActiveGuessIndex((activeIndex) =>
          Math.min(activeIndex + 1, WORD_SIZE - 1)
        );
      } else if (ev.key === "ArrowLeft") {
        setActiveGuessIndex((activeIndex) =>
          activeIndex !== null ? Math.max(activeIndex - 1, 0) : 0
        );
      }
    };
    const moveActiveLetterWhenLetterInput = (ev: KeyboardEvent) => {
      if (!letterRegex.test(ev.key)) return;

      setActiveGuessIndex((activeGuessIndex) =>
        Math.min(activeGuessIndex + 1, WORD_SIZE - 1)
      );
    };

    document.addEventListener("keydown", handleArrowClick);
    document.addEventListener("keydown", moveActiveLetterWhenLetterInput);

    return () => {
      document.removeEventListener("keydown", handleArrowClick);
      document.removeEventListener("keydown", moveActiveLetterWhenLetterInput);
    };
  }, [activeGuessIndex]);

  return (
    <div
      className={`${styles.guessRow} ${active ? styles.active : ""}`}
      role="guess-row"
    >
      {[...Array(WORD_SIZE)].map((_, index) => (
        <div
          onClick={() => {
            setActiveGuessIndex(index);
          }}
          role="guess-wrapper"
          key={index}
        >
          <GuessLetter active={active && activeGuessIndex === index} />
        </div>
      ))}
    </div>
  );
};
