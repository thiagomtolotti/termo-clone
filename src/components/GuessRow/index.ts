import { GuessRow } from "./GuessRow";

export default GuessRow;
