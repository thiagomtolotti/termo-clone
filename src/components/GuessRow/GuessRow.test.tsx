import { fireEvent, render } from "@testing-library/react";
import { GuessRow } from "./GuessRow";

describe("GuessRow element", () => {
  it("Should render correctly", () => {
    const { container } = render(<GuessRow />);

    expect(container).toMatchSnapshot();
  });

  describe("if is active", () => {
    it("Should have at least one active guess", () => {
      const { getAllByRole } = render(<GuessRow active />);

      const guessLetters = getAllByRole("guess");

      const hasActiveGuess = guessLetters.some((guesLetter) =>
        guesLetter.classList.contains("active")
      );

      expect(hasActiveGuess).toBe(true);
    });

    it("Should allow the user to click on the guesses to make it active", () => {
      const { getAllByRole } = render(<GuessRow active />);

      const guessLetters = getAllByRole("guess");

      fireEvent.click(guessLetters[2]);

      expect(guessLetters[2].classList.contains("active")).toBe(true);
      expect(guessLetters[0].classList.contains("active")).toBe(false);
    });

    it("Should allow the user to press the keys to make them active", () => {
      expect(true).toBeFalsy();
    });
  });

  describe("it is not active", () => {
    it("Shouldn't have an active guess", () => {
      const { getAllByRole } = render(<GuessRow />);

      const guessLetters = getAllByRole("guess-wrapper");

      const hasActiveGuess = guessLetters.some((guesLetter) =>
        guesLetter.classList.contains("active")
      );

      expect(hasActiveGuess).toBe(false);
    });
  });
});
