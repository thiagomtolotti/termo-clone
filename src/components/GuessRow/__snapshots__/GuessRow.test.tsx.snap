// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuessRow element Should render correctly 1`] = `
<div>
  <div
    class="guessRow "
    role="guess-row"
  >
    <div
      role="guess-wrapper"
    >
      <div
        class="guess "
        role="guess"
      />
    </div>
    <div
      role="guess-wrapper"
    >
      <div
        class="guess "
        role="guess"
      />
    </div>
    <div
      role="guess-wrapper"
    >
      <div
        class="guess "
        role="guess"
      />
    </div>
    <div
      role="guess-wrapper"
    >
      <div
        class="guess "
        role="guess"
      />
    </div>
    <div
      role="guess-wrapper"
    >
      <div
        class="guess "
        role="guess"
      />
    </div>
  </div>
</div>
`;
